src/librustc_parse/mbe_fragment.rs

```diff
@@ -0,0 +1,200 @@
+use rustc_ast::ptr::P;
+use rustc_ast::token::{self, Nonterminal, Token};
+use rustc_ast_pretty::pprust;
+use rustc_errors::PResult;
+use rustc_span::symbol::{kw, sym, Ident, Symbol};
+
+use crate::parser::{FollowedByType, Parser, PathStyle};
+
+#[derive(Copy, Clone)]
+pub enum MbeFragment {
```

Comment:
FWIW, we already have this enum in `token.rs`, it's called `enum Nonterminal`, except that it has data in addition to tags.
Perhaps these two enums should live together in `rustc_ast` as `Nonterminal` and `NonterminalKind`.

The `may_begin_with` is also very similar to `*can_begin*` functions in `token.rs`.

Moving the nonterminal parsing to `rustc_parse` is certainly a good move.

Reply:
Perhaps it's also a good opportunity to try [`#![feature(arbitrary_enum_discriminant)]`](https://github.com/rust-lang/rfcs/blob/master/text/2363-arbitrary-enum-discriminant.md).